### PR TITLE
Add regex function in string matcher

### PR DESCRIPTION
**Pull request by xulingqing**
Add regex matching for Authz policies:
- code change https://github.com/istio/istio/pull/27984

istio.io website change:
- istio/istio.io    https://github.com/istio/istio.io/pull/8340
- istio/api          https://github.com/istio/api/pull/1707

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/security/authz/model/generator.go [hunk 1]
```diff
@@ -116,7 +116,13 @@ func (srcNamespaceGenerator) permission(_, _ string, _ bool) (*rbacpb.Permission
 }
 
 func (srcNamespaceGenerator) principal(_, value string, forTCP bool) (*rbacpb.Principal, error) {
-	v := strings.Replace(value, "*", ".*", -1)
+	var v string
+	if strings.HasPrefix(value, "regex:") {
```
Comment (howardjohn) [inline pilot/pkg/security/authz/model/generator.go:120]: can you add a test for this case?

File: tests/integration/security/authorization_test.go [hunk 2]
```diff
@@ -1103,6 +1103,77 @@ func TestAuthorization_Path(t *testing.T) {
 		})
 }
 
+func TestAuthorization_Path_Regex(t *testing.T) {
```
Comment (yangminzhu) [inline tests/integration/security/authorization_test.go:1106]: No need for new tests, just update the existing one is enough.

File: pilot/pkg/security/authz/matcher/string.go [hunk 3]
```diff
@@ -48,6 +48,14 @@ func StringMatcherWithPrefix(v, prefix string) *matcherpb.StringMatcher {
 	// the Envoy StringMatcher doesn't allow empty prefix/suffix.
 	case v == "*":
 		return StringMatcherRegex(".+")
+	case strings.HasPrefix(v, "regex:"):
+		return &matcherpb.StringMatcher{
+			MatchPattern: &matcherpb.StringMatcher_SafeRegex{
+				SafeRegex: &matcherpb.RegexMatcher{
+					Regex: strings.TrimPrefix(v, "regex:"),
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/string.go:55]: you need to take care of the `prefix` argument, see how it is handled in other cases.

**Review by yangminzhu — COMMENTED** [review]
You should update the tests under the builder, see all the golden files in testdata.

Also add a new global environment flag to make this behavior opt-in.

**Comment by yangminzhu** [comment]
This is also user-facing changes (even guarded by flag and disabled by default), it needs a feature file.

File: pilot/pkg/security/authz/builder/testdata/action-deny-HTTP-for-TCP-filter-out.yaml [hunk 4]
```diff
@@ -198,21 +198,11 @@ typedConfig:
                       safeRegex:
                         googleRe2: {}
                         regex: .*/ns/ns/.*
-                - authenticated:
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/testdata/action-deny-HTTP-for-TCP-filter-out.yaml:201]: Why are these removed in the output?

Comment (xulingqing) [reply pilot/pkg/security/authz/builder/testdata/action-deny-HTTP-for-TCP-filter-out.yaml:201]: I'm not sure. It's removed after I run `make gen`  I'll recover it in next commit. Thanks!

File: pilot/pkg/security/authz/model/generator.go [hunk 5]
```diff
@@ -116,7 +118,16 @@ func (srcNamespaceGenerator) permission(_, _ string, _ bool) (*rbacpb.Permission
...
 func (srcNamespaceGenerator) principal(_, value string, forTCP bool) (*rbacpb.Principal, error) {
-	v := strings.Replace(value, "*", ".*", -1)
+	var v string
+	if features.EnableAuthzRegexMatching && strings.HasPrefix(value, "regex:") {
+		v = strings.TrimPrefix(value, "regex:")
+	} else {
+		v = strings.Replace(value, "*", ".*", -1)
+	}
+
+	if err := validation.ValidateRegex(v); err != nil {
```
Comment (howardjohn) [inline pilot/pkg/security/authz/model/generator.go:128]: We should not be doing validation on the config generation path. This is extremely expensive. Should validate it in the validation only.

File: tests/integration/security/authorization_test.go [hunk 6]
```diff
@@ -1093,13 +1093,21 @@ func TestAuthorization_Path(t *testing.T) {
...
+				newTestCase("/abc/0", true),
+				newTestCase("/abc/12",  false),
+				newTestCase("/abc/a", false),
+				newTestCase("/abc/0/123", false),
 			}
 
 			args := map[string]string{
 				"Namespace": ns.Name(),
 			}
+			fmt.Println("here main test")
```
Comment (howardjohn) [inline tests/integration/security/authorization_test.go:1107]: remove this

**Review by howardjohn — DISMISSED** [review]

File: pilot/pkg/security/authz/model/generator.go (hunk 5 above)
Comment (xulingqing) [reply pilot/pkg/security/authz/model/generator.go:128]: If we only validate the regex in validation, where should I call the validation.validateRegex(v) ?

File: pilot/pkg/security/authz/builder/testdata/all-fields-out.yaml [hunk 7]
```diff
@@ -65,6 +74,9 @@ typedConfig:
                   - header:
                       name: :method
                       presentMatch: true
+                  - header:
+                      exactMatch: regex:not-method-regex
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/testdata/all-fields-out.yaml:78]: some of these are still using exact matching?

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 8]
```diff
@@ -128,6 +129,7 @@ func TestGenerator_GenerateHTTP(t *testing.T) {
 		},
 	}
 
+	features.EnableAuthzRegexMatching = true
```
[the remark this answers is not in the document]
Comment (yangminzhu) [reply pilot/pkg/security/authz/builder/builder_test.go:132]: Also test the TCP case in addition to the HTTP one.

File: releasenotes/notes/27984.yaml [hunk 9]
```diff
@@ -0,0 +1,11 @@
...
+kind: feature
+area: security
+issue:
+  - 27984
+
+# releaseNotes is a markdown listing of any user facing changes. This will appear in the
+# release notes.
+releaseNotes:
+- |
+  **Added** Enable regex matching for authorization policy
```
Comment (yangminzhu) [inline releasenotes/notes/27984.yaml:11]: Make it clear that the feature is disabled by default, and can be enabled by setting the global flag on istiod.

File: tests/integration/security/testdata/authz/v1beta1-path.yaml.tmpl [hunk 10]
```diff
@@ -14,6 +14,6 @@ spec:
   rules:
   - to:
     - operation:
-        paths: ["/public*"]
+        paths: ["/public*", 'regex:\/abc\/[0-9]?']
```
Comment (yangminzhu) [inline tests/integration/security/testdata/authz/v1beta1-path.yaml.tmpl:17]: let's use `regex` instead of `abc` to make this a bit more obvious that it is testing regex.

Comment (yangminzhu) [reply tests/integration/security/testdata/authz/v1beta1-path.yaml.tmpl:17]: can you also update the following tests:
1. `TestAuthorization_Conditions` to cover the request.headers and source.principal
2. `TestAuthorization_TCP` to cover the source.principal
3. `TestAuthorization_JWT` to cover jwt claim

File: pilot/pkg/security/authz/matcher/string.go [hunk 11]
```diff
@@ -27,6 +29,9 @@ func StringMatcher(v string) *matcherpb.StringMatcher {
 
 // StringMatcherRegex creates a regex string matcher for regex.
 func StringMatcherRegex(regex string) *matcherpb.StringMatcher {
+	if features.EnableAuthzRegexMatching && strings.HasPrefix(regex, "regex:") {
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/string.go:32]: Also update the comment to match the new behavior.

File: pilot/pkg/security/authz/matcher/string.go [hunk 12]
```diff
@@ -43,9 +48,17 @@ func StringMatcherRegex(regex string) *matcherpb.StringMatcher {
 // created string matcher, note the prefix is ignored if v is wildcard ("*").
 // The wildcard "*" will be generated as ".+" instead of ".*".
 func StringMatcherWithPrefix(v, prefix string) *matcherpb.StringMatcher {
+	// If environment variable AUTHZ_ENABLE_REGEX is set to ture, prefix "regex:"
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/string.go:51]: Also update the function comment to match the new behavior.

Comment (yangminzhu) [reply pilot/pkg/security/authz/matcher/string.go:51]: also typo in `triger`

File: pilot/pkg/security/authz/matcher/string.go (hunk 11 above)
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/string.go:32]: Do we need to check the flag again in StringMatcherRegex? It seems it's already enough that this is already done in StringMatcherWithPrefix and `func (srcNamespaceGenerator) principal`.

File: pilot/pkg/security/authz/matcher/string_test.go [hunk 13]
```diff
@@ -30,6 +32,7 @@ type testCase struct {
 }
 
 func TestStringMatcherWithPrefix(t *testing.T) {
+	features.EnableAuthzRegexMatching = true
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/string_test.go:35]: Do not set the flag to true for all tests, just set it to specific test case to make sure we cover both true and false cases.

Comment (xulingqing) [reply pilot/pkg/security/authz/matcher/string_test.go]: Done adding tests for both enabled and disabled case to StringMatcher and HeaderMatcher.

File: pilot/pkg/security/authz/matcher/string.go (hunk 11 above)
Comment (xulingqing) [reply pilot/pkg/security/authz/matcher/string.go]: Yes, thanks! removed.

File: releasenotes/notes/27984.yaml (hunk 9 above)
Comment (xulingqing) [reply releasenotes/notes/27984.yaml]: modified. Thanks. Let me know if you would like to change the content.

File: pilot/pkg/security/authz/builder/testdata/all-fields-out.yaml (hunk 7 above)
Comment (xulingqing) [reply pilot/pkg/security/authz/builder/testdata/all-fields-out.yaml]: Fixed HeaderMatching. Thanks!

File: pilot/pkg/security/authz/matcher/string.go (hunk 12 above)
Comment (xulingqing) [reply pilot/pkg/security/authz/matcher/string.go:48]: Do you think the current comment explains well? Thanks.

File: pilot/pkg/security/authz/builder/builder_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (xulingqing) [reply pilot/pkg/security/authz/builder/builder_test.go]: Do you prefer to use same input or different input for these integration testing? Now I'm using same input and test for both enabled and disabled conditions for regex matching.

File: pilot/pkg/security/authz/matcher/header.go [hunk 14]
```diff
@@ -19,10 +19,28 @@ import (
...
+	// will triger regex matching for Authz policy filed header.
+	// Those values that not start with "regex:" will be considered as prefix, suffix,
+	// present or exact match based on matching schemas
+	if features.EnableAuthzRegexMatching && strings.HasPrefix(v, "regex:") {
+		v = strings.TrimPrefix(v, "regex:")
+		return &routepb.HeaderMatcher{
+			Name: k,
+			HeaderMatchSpecifier: &routepb.HeaderMatcher_SafeRegexMatch{
+				SafeRegexMatch: &matcherpb.RegexMatcher{
+					Regex: v,
```
Comment (howardjohn) [inline pilot/pkg/security/authz/matcher/header.go:38]: I thought `EngineType` was required?

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl [hunk 15]
```diff
@@ -160,28 +134,22 @@ spec:
   rules:
   - to:
     - operation:
-        paths: ["/source-principal-a"]
+        paths: ["/source-principal-a", 'regex:\/source-principal-a-regexp\/[0-9]?']
     when:
     - key: source.principal
       values: ["cluster.local/ns/{{ .NamespaceA }}/sa/a"]
+- to:
```
Comment (yangminzhu) [inline tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:141]: wrong indentation

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl [hunk 16]
```diff
@@ -118,28 +98,22 @@ spec:
...
+      values: ["regex:{{ .NamespaceA }}"]
   - to:
     - operation:
         paths: ["/source-namespace-b"]
     when:
     - key: source.namespace
       values: ["{{ .NamespaceB }}"]
----
-
-apiVersion: security.istio.io/v1beta1
```
Comment (yangminzhu) [inline tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:133]: why are you removing this and merging them to a single policy? you can just add your regex to existing policy or better just add a new policy for regex.

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 17]
```diff
@@ -51,19 +52,22 @@ func TestGenerator_GenerateHTTP(t *testing.T) {
...
 		isVersion14 bool
+		enableRegex bool
 		input       string
 		want        []string
 	}{
 		{
 			name:        "path14",
 			input:       "path14-in.yaml",
 			isVersion14: true,
+			enableRegex: true,
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/builder_test.go:63]: Don't need regex for path14, same for path15, it should be covered in all-fields-regex-enabled

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 18]
```diff
@@ -142,33 +154,55 @@ func TestGenerator_GenerateHTTP(t *testing.T) {
 			verify(t, convertHTTP(got), tc.want, false /* forTCP */)
 		})
 	}
+
+	features.EnableAuthzRegexMatching = false
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/builder_test.go:158]: put this inside the t.Run() above

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 19]
```diff
@@ -129,6 +140,7 @@ func TestGenerator_GenerateHTTP(t *testing.T) {
 	}
 
 	for _, tc := range testCases {
+		features.EnableAuthzRegexMatching = tc.enableRegex
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/builder_test.go:143]: Put this inside t.Run below

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 20]
```diff
@@ -142,33 +154,55 @@ func TestGenerator_GenerateHTTP(t *testing.T) {
...
 		},
 		{
-			name:  "action-deny-HTTP-for-TCP-filter",
+			name:        "action-allow-HTTP-for-TCP-filter-regex-enabled",
+			input:       "action-allow-HTTP-for-TCP-filter-in.yaml",
+			enableRegex: true,
+			want:        []string{"action-allow-HTTP-for-TCP-filter-regex-out.yaml"},
+		},
+		{
+			name:  "action-deny-HTTP-for-TCP-filter-regex-enabled",
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/builder_test.go:181]: wrong name, you didn't enabled regex for this test, and also below.

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 21]
```diff
@@ -142,33 +154,55 @@ func TestGenerator_GenerateHTTP(t *testing.T) {
...
+		{
+			name:        "action-audit-HTTP-for-TCP-filter-regex-enabled",
+			input:       "action-audit-HTTP-for-TCP-filter-in.yaml",
+			enableRegex: true,
+			want:        []string{"action-audit-HTTP-for-TCP-filter-regex-out.yaml"},
+		},
 	}
 
 	for _, tc := range testCases {
+		features.EnableAuthzRegexMatching = tc.enableRegex
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/builder_test.go:205]: same here, put it inside t.Run

File: pilot/pkg/security/authz/builder/builder_test.go [hunk 22]
```diff
@@ -182,6 +216,7 @@ func TestGenerator_GenerateTCP(t *testing.T) {
 			verify(t, convertTCP(got), tc.want, true /* forTCP */)
 		})
 	}
+	features.EnableAuthzRegexMatching = false
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/builder/builder_test.go:219]: ditto.

File: pilot/pkg/features/pilot.go [hunk 23]
```diff
@@ -400,4 +400,7 @@ var (
 		15*time.Second,
 		"If set, the max amount of time to delay a push by. Depends on PILOT_ENABLE_FLOW_CONTROL.",
 	).Get()
+
+	EnableAuthzRegexMatching = env.RegisterBoolVar("AUTHZ_ENABLE_REGEX", false,
+		"If true, regex matching will be enabled for authorization policy. In the policy, regex expression after 'regex:' will be used").Get()
```
Comment (yangminzhu) [inline pilot/pkg/features/pilot.go:405]: `In the policy, regex expression after 'regex:' will be used` -> `In the policy, text after `regex:` will be interpreted as regular expression`

File: pilot/pkg/security/authz/matcher/header.go [hunk 24]
```diff
@@ -19,10 +19,28 @@ import (
 
 	routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
 	matcherpb "github.com/envoyproxy/go-control-plane/envoy/type/matcher/v3"
+
+	"istio.io/istio/pilot/pkg/features"
 )
 
 // HeaderMatcher converts a key, value string pair to a corresponding HeaderMatcher.
 func HeaderMatcher(k, v string) *routepb.HeaderMatcher {
+	// If environment variable AUTHZ_ENABLE_REGEX is set to ture, prefix "regex:"
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/header.go:28]: I think you can just remove these comments, the code is pretty clear about this.

File: pilot/pkg/security/authz/matcher/header_test.go [hunk 25]
```diff
@@ -52,14 +55,55 @@ func TestHeaderMatcher(t *testing.T) {
...
+			V:           `regex:\/productpage\/[0-9]?`,
+			enableRegex: false,
+			Expect: &routepb.HeaderMatcher{
+				Name: ":path",
+				HeaderMatchSpecifier: &routepb.HeaderMatcher_ExactMatch{
+					ExactMatch: `regex:\/productpage\/[0-9]?`,
+				},
+			},
+		},
+		{
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/header_test.go:70]: this one can be removed, it's enough with the above one.

File: pilot/pkg/security/authz/matcher/header_test.go [hunk 26]
```diff
@@ -52,14 +55,55 @@ func TestHeaderMatcher(t *testing.T) {
 				},
 			},
 		},
+		{
+			Name:        "regex match - disabled - ExactMatch",
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/header_test.go:59]: just change the name to `regex match disabled`, and the other one to `regex match enabled`

File: pilot/pkg/security/authz/matcher/header_test.go [hunk 27]
```diff
@@ -52,14 +55,55 @@ func TestHeaderMatcher(t *testing.T) {
...
 	}
 
 	for _, tc := range testCases {
+		features.EnableAuthzRegexMatching = tc.enableRegex
 		actual := HeaderMatcher(tc.K, tc.V)
 		if !cmp.Equal(tc.Expect, actual, protocmp.Transform()) {
 			t.Errorf("expecting %v, but got %v", tc.Expect, actual)
 		}
 	}
+
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/header_test.go:105]: you can rewrite the current code to use t.Run() and then put this inside it.

File: pilot/pkg/security/authz/matcher/string.go (hunk 12 above)
Comment (yangminzhu) [reply pilot/pkg/security/authz/matcher/string.go:48]: I mean update the function comment above in line 44 (just mention `regex:` will be interpreted as regular expression when the flag is enabled), not here. Actually you can remove the comments here.

File: pilot/pkg/security/authz/matcher/string.go [hunk 28]
```diff
@@ -43,9 +45,19 @@ func StringMatcherRegex(regex string) *matcherpb.StringMatcher {
...
+	// Those values that not start with "regex:" will be considered as prefix, suffix,
+	// present or exact match based on matching schemas
+	if features.EnableAuthzRegexMatching && strings.HasPrefix(v, "regex:") {
+		return StringMatcherRegex(prefix + strings.TrimPrefix(v, "regex:"))
+	}
+
 	switch {
 	// Check if v is "*" first to make sure we won't generate an empty prefix/suffix StringMatcher,
 	// the Envoy StringMatcher doesn't allow empty prefix/suffix.
+	// If v starts with "regex:" but the system doesn't EnableAuthzRegexMatching,
```
Comment (yangminzhu) [inline pilot/pkg/security/authz/matcher/string.go:59]: just remove the comment, redundant.

File: pkg/config/security/security.go [hunk 29]
```diff
@@ -181,3 +191,16 @@ func validateMapKey(key string) error {
 	}
 	return fmt.Errorf("bad key (%s): should have format a[b]", key)
 }
+
+func validateRegex(values []string) error {
+	if features.EnableAuthzRegexMatching {
+		for _, v := range values {
+			if strings.HasPrefix(v, "regex:") {
+				if _, err := regexp.Compile(strings.TrimPrefix(v, "regex:")); err != nil {
+					return fmt.Errorf("%s error while compiling %s", err, v)
```
Comment (yangminzhu) [inline pkg/config/security/security.go:200]: just return err.

File: tests/integration/security/authorization_test.go [hunk 30]
```diff
@@ -862,9 +862,12 @@ func TestAuthorization_TCP(t *testing.T) {
 				// - request to port http-8090 should be denied because both path and port are matched.
 				// - request to port http-8091 should be allowed because the port is not matched.
 				// - request to port tcp-8092 should be allowed because the port is not matched.
-				newTestCase(a, b, "http-8090", false, scheme.HTTP),
-				newTestCase(a, b, "http-8091", true, scheme.HTTP),
-				newTestCase(a, b, "tcp-8092", true, scheme.TCP),
+				newTestCase(a, b, "/data", "http-8090", false, scheme.HTTP),
+				newTestCase(a, b, "/data", "http-8091", true, scheme.HTTP),
+				newTestCase(a, b, "/data", "tcp-8092", true, scheme.TCP),
+				newTestCase(a, b, "/regexp/", "http-8090", false, scheme.HTTP),
```
Comment (yangminzhu) [inline tests/integration/security/authorization_test.go:868]: no, we shouldn't use path in the TestAuthorization_TCP test, the /data before is just an placeholder and not used in the authz policy, it might work here as this specific test case is using HTTP port but that is not the purpose of this test (HTTP is covered in other tests). 

To test regex in TCP, just use those attributes that are available in TCP, e.g. source principal or namespace, and to distinguish with other tests you can create new app.

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl (hunk 16 above)
Comment (xulingqing) [reply tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:133]: It seems simplified the yaml file. Do you prefer I keep original policy or merge them together?

**Comment by yangminzhu** [comment]
I prefer to keep original policy, you can add the regex to existing policy
or in a new one.

File: pilot/pkg/security/authz/builder/builder_test.go (hunk 17 above)
Comment (xulingqing) [reply pilot/pkg/security/authz/builder/builder_test.go:63]: But "all fields in" are only testing regex matching after version 15, do we need to keep it for path-14?

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl (hunk 16 above)
Comment (xulingqing) [reply tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:133]: Sure I will recover to original policy. Thanks!

File: tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl [hunk 31]
```diff
@@ -42,14 +42,20 @@ spec:
...
+      values: ["regex:foo-regexp[0-9]?"]
 ---
 
 # Each of the following authorization policy uses a different condition on the given path.
 
 apiVersion: security.istio.io/v1beta1
 kind: AuthorizationPolicy
 metadata:
-  name: condition-request-headers-notvalues
+  name: condition-request-headers
```
Comment (yangminzhu) [inline tests/integration/security/testdata/authz/v1beta1-conditions.yaml.tmpl:58]: why change the name here? It's duplicate of the other policy and will override it.

**Comment by WesleySensedia** [comment]
@xulingqing xulingqing and @howardjohn. Any chance to bring this to life?
